File: java/src/main/java/com/workshop/architecture/fitness/clean_architecture/adapter/gateway/jpa/JpaMembershipRepository.java
package com.workshop.architecture.fitness.clean_architecture.adapter.gateway.jpa;

import com.workshop.architecture.fitness.clean_architecture.entity.CustomerId;
import com.workshop.architecture.fitness.clean_architecture.entity.Duration;
import com.workshop.architecture.fitness.clean_architecture.entity.Membership;
import com.workshop.architecture.fitness.clean_architecture.entity.MembershipId;
import com.workshop.architecture.fitness.clean_architecture.entity.MembershipStatus;
import com.workshop.architecture.fitness.clean_architecture.entity.MembershipStatusValue;
import com.workshop.architecture.fitness.clean_architecture.entity.PlanDetails;
import com.workshop.architecture.fitness.clean_architecture.entity.PlanId;
import com.workshop.architecture.fitness.clean_architecture.entity.Price;
import com.workshop.architecture.fitness.clean_architecture.use_case.port.outbound.ForStoringMemberships;
import com.workshop.architecture.fitness.layered.infrastructure.MembershipEntity;
import com.workshop.architecture.fitness.layered.infrastructure.MembershipRepository;
import java.util.UUID;
import org.springframework.stereotype.Component;

@Component
public class JpaMembershipRepository implements ForStoringMemberships {
    private final MembershipRepository membershipRepository;

    public JpaMembershipRepository(MembershipRepository membershipRepository) {
        this.membershipRepository = membershipRepository;
    }

    @Override
    public Membership storeMembership(Membership membership) {
        var entity = new MembershipEntity(
                membership.id(),
                membership.customerId().toString(),
                membership.planId().toString(),
                membership.planPrice(),
                membership.planDurationInMonths(),
                membership.status(),
                membership.statusReason(),
                membership.startDate(),
                membership.endDate()
        );

        var storedEntity = membershipRepository.save(entity);

        return new Membership(
                new MembershipId(storedEntity.getId()),
                new CustomerId(UUID.fromString(storedEntity.getCustomerId())),
                new MembershipStatus(
                        MembershipStatusValue.valueOf(storedEntity.getStatus().toUpperCase()),
                        storedEntity.getReason()
                ),
                new PlanDetails(
                        new PlanId(UUID.fromString(storedEntity.getPlanId())),
                        new Price(storedEntity.getPlanPrice()),
                        new Duration(storedEntity.getStartDate(), storedEntity.getEndDate())
                )
        );
    }
}

File: python/src/workshop_api/fitness/hexagon/outside/driven/repository/jpa/jpa_membership_repository.py
import uuid

from workshop_api.fitness.hexagon.inside.port.outbound.for_storing_memberships import (
    ForStoringMemberships,
)
from workshop_api.fitness.hexagon.inside.port.outbound.membership import Membership
from workshop_api.fitness.layered.infrastructure.membership_entity import MembershipOrmModel
from workshop_api.fitness.layered.infrastructure.membership_repository import (
    MembershipRepository,
)


class JpaMembershipRepository(ForStoringMemberships):
    def __init__(self, membership_repository: MembershipRepository) -> None:
        self.membership_repository = membership_repository

    def store_membership(self, membership: Membership) -> Membership:
        entity = MembershipOrmModel(
            id=str(membership.id),
            customer_id=str(membership.customer_id),
            plan_id=str(membership.plan_id),
            plan_price=membership.plan_price,
            plan_duration=membership.plan_duration_in_months,
            status=membership.status,
            reason=membership.status_reason,
            start_date=membership.start_date,
            end_date=membership.end_date,
        )

        stored_entity = self.membership_repository.save(entity)

        return Membership(
            id=uuid.UUID(stored_entity.id),
            customer_id=uuid.UUID(stored_entity.customer_id),
            plan_id=uuid.UUID(stored_entity.plan_id),
            plan_price=stored_entity.plan_price,
            plan_duration_in_months=stored_entity.plan_duration,
            status=stored_entity.status,
            status_reason=stored_entity.reason,
            start_date=stored_entity.start_date,
            end_date=stored_entity.end_date,
        )
